kimchi/src/circuits/polynomials/foreign_field_add/circuitgates.rs

```diff
@@ -0,0 +1,233 @@
+pub struct ForeignFieldFin<F>(PhantomData<F>);
+
+impl<F> Argument<F> for ForeignFieldFin<F>
+where
+    F: FftField,
+{
+    const ARGUMENT_TYPE: ArgumentType = ArgumentType::Gate(GateType::ForeignFieldFin);
+    const CONSTRAINTS: u32 = 5;
+
+    fn constraint_checks<T: ExprOps<F>>(env: &ArgumentEnv<F, T>) -> Vec<T> {
+        // Similar to performing a foreign field addition gate with these inputs
+        // left_input_lo -> resmin_lo  right_input_lo -> 0     result_lo -> bound_lo  carry_lo -> bound_carry_lo
+        // left_input_mi -> resmin_mi  right_input_mi -> 0     result_mi -> bound_mi  carry_mi -> bound_carry_mi
+        // left_input_hi -> resmin_hi  right_input_hi -> 2^88  result_hi -> bound_hi
+        // field_overflow -> 1         sign -> 1
+
+        // check addition for bound was performed correctly
+
+        let resmin_lo = env.witness_curr(0);
+        let resmin_mi = env.witness_curr(1);
+        let resmin_hi = env.witness_curr(2);
+
+        let bound_lo = env.witness_curr(3);
+        let bound_mi = env.witness_curr(4);
+        let bound_hi = env.witness_curr(5);
+
+        let carry_lo = env.witness_curr(7);
+        let carry_mi = env.witness_curr(8);
+
+        let modulus: [T; LIMB_COUNT] = array::from_fn(|i| env.foreign_modulus(i));
+        let two_to_limb = T::literal(F::from(2u128.pow(LIMB_BITS as u32)));
+
+        // check values of carry bits
+        let mut checks = carry(env);
+
+        let computed_bound_lo =
+            resmin_lo - modulus[0].clone() - carry_lo.clone() * two_to_limb.clone();
+        let computed_bound_mi =
+            resmin_mi - modulus[1].clone() - carry_mi.clone() * two_to_limb.clone() + carry_lo;
+        let computed_bound_hi = resmin_hi - modulus[2].clone() + carry_mi + two_to_limb;
+
+        checks.push(bound_lo - computed_bound_lo);
+        checks.push(bound_mi - computed_bound_mi);
+        checks.push(bound_hi - computed_bound_hi);
+        checks
+    }
+}
+
+/// Auxiliary function to obtain the constraints of a foreign field addition result
+fn result<F: FftField, T: ExprOps<F>>(env: &ArgumentEnv<F, T>) -> Vec<T> {
+    let foreign_modulus: [T; LIMB_COUNT] = array::from_fn(|i| env.foreign_modulus(i));
+
+    let two_to_limb = T::literal(F::from(2u128.pow(LIMB_BITS as u32)));
+
+    let left_input_lo = env.witness_curr(0);
+    let left_input_mi = env.witness_curr(1);
+    let left_input_hi = env.witness_curr(2);
+
+    let right_input_lo = env.witness_curr(3);
+    let right_input_mi = env.witness_curr(4);
+    let right_input_hi = env.witness_curr(5);
+
+    let field_overflow = env.witness_curr(6);
+
+    // Result carry bits for limb overflows / underflows.
+    let carry_lo = env.witness_curr(7);
+    let carry_mi = env.witness_curr(8);
+
+    let sign = env.witness_curr(9);
+
+    let result_lo = env.witness_next(0);
+    let result_mi = env.witness_next(1);
+    let result_hi = env.witness_next(2);
+
+    // r_0 = a_0 + s * b_0 - q * f_0 - 2^88 * c_0
```

Comment:
These three lgtm!

Reply:
Nice, here you can see the missing r_0, r_1, r_2 from the previous comment 😉